source/slang/slang-lower-to-ir.cpp

```diff
@@ -6470,7 +6547,53 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
         // body appear before the function itself in the list
         // of global values.
         irFunc->moveToEnd();
-        return LoweredValInfo::simple(finishOuterGenerics(subBuilder, irFunc));
+
+        // If this function is defined inside an interface, add a reference to the IRFunc from
+        // the interface's type definition.
+        auto finalVal = finishOuterGenerics(subBuilder, irFunc);
+
+        if (auto genericVal = as<IRGeneric>(finalVal))
+        {
+            auto funcType = lowerFuncType(decl);
```

Comment:
For generic functions I'll have to run the `lowerFuncType` logic all over again. This is unfortunate but I can't come up with a better way.

Reply:
It seems like a general-purpose fix might be to make `finishOuterGenerics` handle the problem of giving a type to each outer `IRGeneric` using the following approach:

* Take the type of the inner instruction (the thing you are going to `return` from the generic
* Construct a new `IRGeneric` that will represent the type of this generic
* Clone all the parameters of this `IRGeneric` into the new one, and build up an `IRCloneEnv` that maps the original parameters over to the new one
* Now clone the type of the inner instruction in that newly-construct `IRCloneEnv` and make it the return value of our new `IRGeneric`
* Now we can make the type of the original `IRGeneric` be the new `IRGeneric`
* TODO: figure out what type to give to the new `IRGeneric` so that this doesn't result in infinite regress.

Reply:
Oh! One thing I just realized is an issue with using an `IRGeneric` to represent a `forall` type to be used as the type of another generic is that we can't deduplicate such types using their operands, so we'd have no way to easily and robustly answer questions like "do these two generic functions have the same (`forall`-based) type?"

That issue might be reason enough to start thinking about how we can represent `forall` types more directly. In the worst case I believe we can do something using [De Bruijin indices](https://en.wikipedia.org/wiki/De_Bruijn_index).

Reply:
Will address in follow-up PRs.